packages/crypto/src/Bip32/Bip32PrivateKey.ts

```diff
@@ -0,0 +1,174 @@
+/* eslint-disable  no-bitwise */
+import * as Bip32KeyDerivation from './Bip32KeyDerivation';
+import { Bip32PublicKey } from './Bip32PublicKey';
+import { EXTENDED_ED25519_PRIVATE_KEY_LENGTH, Ed25519PrivateKey, NORMAL_ED25519_PRIVATE_KEY_LENGTH } from '../Ed25519e';
+import { InvalidArgumentError } from '../errors';
+import { crypto_scalarmult_ed25519_base_noclamp } from 'libsodium-wrappers-sumo';
+import { pbkdf2 } from 'pbkdf2';
+
+const SCALAR_INDEX = 0;
+const SCALAR_SIZE = 32;
+const PBKDF2_ITERATIONS = 4096;
+const PBKDF2_KEY_SIZE = 96;
+const PBKDF2_DIGEST_ALGORITHM = 'sha512';
+const CHAIN_CODE_INDEX = 64;
+const CHAIN_CODE_SIZE = 32;
+
+/**
+ * clamp the scalar by:
+ *
+ *  1. clearing the 3 lower bits.
+ *  2. clearing the highest bit.
+ *  3. setting the second-highest bit.
+ *
+ * @param scalar The clamped scalar.
+ */
+const clampScalar = (scalar: Buffer): Buffer => {
+  scalar[0] &= 0b1111_1000;
+  scalar[31] &= 0b0011_1111;
+  scalar[31] |= 0b0100_0000;
+  return scalar;
+};
+
+/**
+ * Extract the scalar part (first 32 bytes) from the extended key.
+ *
+ * @param extendedKey The extended key.
+ * @returns the scalar part of the extended key.
+ */
+const extendedScalar = (extendedKey: Uint8Array) => extendedKey.slice(SCALAR_INDEX, SCALAR_SIZE);
+
+export const BIP32_ED25519_PRIVATE_KEY_LENGTH = 96;
+
+/**
+ * Bip32PrivateKey private key. This type of key have the ability to derive additional keys from them
+ * following the BIP32 derivation scheme variant called BIP32-Ed25519.
+ *
+ * @see <a href="https://ieeexplore.ieee.org/document/7966967">
+ *        BIP32-Ed25519: Hierarchical Deterministic Keys over a Non-linear Keyspace
+ *      </a>
+ */
+export class Bip32PrivateKey {
```

Comment:
I think we should keep the original types as OpaqueString (just move them to `crypto` package), because they are being used as string and we still need a type for that.

Classes that provides operations using the key could have a different name and a constructor that takes in the opaque hex string.

Reply:
Not really, OpaqueString is actually just String at runtime, which is very handy for serialization/deserialization

Reply:
For example, data returned from http providers is serialized in cardano-services and deserialized in cardano-services-client. Having those types as classes would make this serialization and deserialization more complicated. That's not relevant for private key types, but relevant for public keys, hashes and signatures.

Also just having a string type that is not interchangeable with other strings is very handy, because you can be sure that you have the correct type of key/hash etc., and can't use another hex string in place of it. An example of that would be that tsc would complain if you tried using private key in place of a public key.

Reply:
Also have in mind that keys, hashes and signatures are almost always passed around as hex strings

Reply:
I think we should keep `core` types as plain javascript objects. It'll keep it more interoperable and you won't need to make any changes to serialization functions that are currently used. What do you think @rhyslbw ?